File: plugins/plugin-slack/python/elizaos_plugin_slack/actions/get_user_info.py
"""
Get user info action for Slack.
"""

from typing import Any, Dict, Optional

from ..types import SLACK_SERVICE_NAME, is_valid_user_id, get_slack_user_display_name


async def validate(runtime: Any, message: Any, state: Optional[Any] = None) -> bool:
    """Validate if this action can be executed."""
    return message.content.get("source") == "slack"


async def handler(
    runtime: Any,
    message: Any,
    state: Optional[Any] = None,
    options: Optional[Dict] = None,
    callback: Optional[Any] = None,
) -> Optional[Dict[str, Any]]:
    """Handle the get user info action."""
    slack_service = runtime.get_service(SLACK_SERVICE_NAME)
    
    if not slack_service or not slack_service.client:
        if callback:
            await callback({"text": "Slack service is not available.", "source": "slack"})
        return {"success": False, "error": "Slack service not available"}
    
    user_id = options.get("user_id") if options else None
    
    if not user_id:
        if callback:
            await callback({
                "text": "Please specify a user ID.",
                "source": "slack",
            })
        return {"success": False, "error": "Missing user_id"}
    
    if not is_valid_user_id(user_id):
        if callback:
            await callback({
                "text": "Invalid user ID format. Slack user IDs start with U.",
                "source": "slack",
            })
        return {"success": False, "error": "Invalid user ID format"}
    
    user = await slack_service.get_user(user_id)
    
    if not user:
        if callback:
            await callback({
                "text": f"Could not find user with ID {user_id}.",
                "source": "slack",
            })
        return {"success": False, "error": "User not found"}
    
    display_name = get_slack_user_display_name(user)
    
    roles = []
    if user.is_admin:
        roles.append("Admin")
    if user.is_owner:
        roles.append("Owner")
    if user.is_primary_owner:
        roles.append("Primary Owner")
    if user.is_bot:
        roles.append("Bot")
    if user.is_restricted:
        roles.append("Guest")
    
    details = [f"**Name:** {display_name}"]
    if user.profile.real_name and user.profile.real_name != display_name:
        details.append(f"**Real Name:** {user.profile.real_name}")
    details.append(f"**Username:** @{user.name}")
    if user.profile.title:
        details.append(f"**Title:** {user.profile.title}")
    if user.profile.email:
        details.append(f"**Email:** {user.profile.email}")
    if user.tz:
        details.append(f"**Timezone:** {user.tz_label or user.tz}")
    if user.profile.status_text:
        emoji = user.profile.status_emoji or ""
        details.append(f"**Status:** {emoji} {user.profile.status_text}")
    if roles:
        details.append(f"**Roles:** {', '.join(roles)}")
    
    if callback:
        await callback({
            "text": f"User information for {display_name}:\n\n" + "\n".join(details),
            "source": "slack",
        })
    
    return {
        "success": True,
        "data": {
            "user_id": user.id,
            "name": user.name,
            "display_name": display_name,
            "real_name": user.profile.real_name,
            "title": user.profile.title,
            "email": user.profile.email,
            "timezone": user.tz,
            "is_admin": user.is_admin,
            "is_owner": user.is_owner,
            "is_bot": user.is_bot,
            "status_text": user.profile.status_text,
            "status_emoji": user.profile.status_emoji,
            "avatar": user.profile.image_192 or user.profile.image_72,
        },
    }


get_user_info = {
    "name": "SLACK_GET_USER_INFO",
    "similes": ["GET_SLACK_USER", "USER_INFO", "SLACK_USER", "WHO_IS"],
    "description": "Get information about a Slack user",
    "validate": validate,
    "handler": handler,
    "examples": [
        [
            {"name": "{{user1}}", "content": {"text": "Who is U0123456789?"}},
            {"name": "{{agent}}", "content": {"text": "Let me look up that user.", "actions": ["SLACK_GET_USER_INFO"]}},
        ],
    ],
}

File: plugins/plugin-discord/actions/getUserInfo.ts
import {
	type Action,
	type ActionExample,
	type ActionResult,
	type Content,
	composePromptFromState,
	type HandlerCallback,
	type HandlerOptions,
	type IAgentRuntime,
	type Memory,
	ModelType,
	type State,
} from "@elizaos/core";
import type { GuildMember } from "discord.js";
import { DISCORD_SERVICE_NAME } from "../constants";
// Import generated prompts
import { getUserInfoTemplate } from "../generated/prompts/typescript/prompts.js";
import { requireActionSpec } from "../generated/specs/spec-helpers";
import type { DiscordService } from "../service";
import { getActionParameters, parseJsonObjectFromText } from "../utils";

const getUserIdentifier = async (
	runtime: IAgentRuntime,
	_message: Memory,
	state: State,
	options?: HandlerOptions,
): Promise<{
	userIdentifier: string;
	detailed: boolean;
} | null> => {
	const parameters = getActionParameters(options);
	if (parameters.userIdentifier) {
		return {
			userIdentifier: String(parameters.userIdentifier),
			detailed:
				parameters.detailed === true ||
				String(parameters.detailed).toLowerCase() === "true",
		};
	}

	const prompt = composePromptFromState({
		state,
		template: getUserInfoTemplate,
	});

	for (let i = 0; i < 3; i++) {
		const response = await runtime.useModel(ModelType.TEXT_SMALL, {
			prompt,
		});

		const parsedResponse = parseJsonObjectFromText(response);
		if (parsedResponse?.userIdentifier) {
			return {
				userIdentifier: String(parsedResponse.userIdentifier),
				detailed:
					parsedResponse.detailed === true ||
					String(parsedResponse.detailed).toLowerCase() === "true",
			};
		}
	}
	return null;
};

const formatUserInfo = (
	member: GuildMember,
	detailed: boolean = false,
): string => {
	const user = member.user;
	const joinedAt = member.joinedAt
		? new Date(member.joinedAt).toLocaleDateString()
		: "Unknown";
	const createdAt = new Date(user.createdAt).toLocaleDateString();
	const roles =
		member.roles.cache
			.filter((role) => role.name !== "@everyone")
			.map((role) => role.name)
			.join(", ") || "No roles";

	const basicInfo = [
		"👤 **User Information**",
		`**Username:** ${user.username}${user.discriminator !== "0" ? `#${user.discriminator}` : ""}`,
		`**Display Name:** ${member.displayName}`,
		`**ID:** ${user.id}`,
		`**Bot:** ${user.bot ? "Yes" : "No"}`,
		`**Account Created:** ${createdAt}`,
	];

	if (detailed) {
		const serverInfo = [
			"",
			"🏛️ **Server Information**",
			`**Nickname:** ${member.nickname || "None"}`,
			`**Joined Server:** ${joinedAt}`,
			`**Roles:** ${roles}`,
			`**Highest Role:** ${member.roles.highest.name}`,
			`**Permissions:** ${member.permissions.toArray().slice(0, 5).join(", ")}${member.permissions.toArray().length > 5 ? "..." : ""}`,
			`**Voice Channel:** ${member.voice.channel ? member.voice.channel.name : "Not in voice"}`,
			`**Status:** ${member.presence?.status || "offline"}`,
		];
		return [...basicInfo, ...serverInfo].join("\n");
	}

	return basicInfo.join("\n");
};

const spec = requireActionSpec("DISCORD_GET_USER_INFO");

export const getUserInfo: Action = {
	name: spec.name,
	similes: spec.similes ? [...spec.similes] : [],
	description: spec.description,
	descriptionCompressed: spec.descriptionCompressed,
	contexts: ["messaging", "connectors"],
	contextGate: { anyOf: ["messaging", "connectors"] },
	roleGate: { minRole: "USER" },
	parameters: [
		{
			name: "userIdentifier",
			description: "Discord user id, mention, username, or current user.",
			required: false,
			schema: { type: "string", default: "current" },
		},
	],
	validate: async (
		runtime: any,
		message: any,
		state?: any,
		options?: any,
	): Promise<boolean> => {
		const __avTextRaw =
			typeof message?.content?.text === "string" ? message.content.text : "";
		const __avText = __avTextRaw.toLowerCase();
		const __avKeywords = ["get", "user", "info"];
		const __avKeywordOk =
			__avKeywords.length > 0 &&
			__avKeywords.some((word) => word.length > 0 && __avText.includes(word));
		const __avRegex = /\b(?:get|user|info)\b/i;
		const __avRegexOk = __avRegex.test(__avText);
		const __avSource = String(message?.content?.source ?? "");
		const __avExpectedSource = "";
		const __avSourceOk = __avExpectedSource
			? __avSource === __avExpectedSource
			: Boolean(
					__avSource ||
						state ||
						runtime?.agentId ||
						runtime?.getService ||
						runtime?.getSetting,
				);
		const __avOptions = options && typeof options === "object" ? options : {};
		const __avInputOk =
			__avText.trim().length > 0 ||
			Object.keys(__avOptions as Record<string, unknown>).length > 0 ||
			Boolean(message?.content && typeof message.content === "object");

		if (!(__avKeywordOk && __avRegexOk && __avSourceOk && __avInputOk)) {
			return false;
		}

		const __avLegacyValidate = async (
			_runtime: IAgentRuntime,
			message: Memory,
			_state?: State,
		): Promise<boolean> => {
			return message.content.source === "discord";
		};
		try {
			return Boolean(
				await (__avLegacyValidate as any)(runtime, message, state, options),
			);
		} catch {
			return false;
		}
	},
	handler: async (
		runtime: IAgentRuntime,
		message: Memory,
		state?: State,
		options?: HandlerOptions,
		callback?: HandlerCallback,
	): Promise<ActionResult | undefined> => {
		const discordService = runtime.getService(
			DISCORD_SERVICE_NAME,
		) as DiscordService;

		if (!discordService?.client) {
			if (callback) {
				await callback?.({
					text: "Discord service is not available.",
					source: "discord",
				});
			}
			return { success: false, error: "Discord service is not available" };
		}

		if (!state) {
			if (callback) {
				await callback?.({
					text: "State is not available.",
					source: "discord",
				});
			}
			return { success: false, error: "State is not available" };
		}

		const userInfo = await getUserIdentifier(runtime, message, state, options);
		if (!userInfo) {
			if (callback) {
				await callback?.({
					text: "I couldn't understand which user you want information about. Please specify a username or mention.",
					source: "discord",
				});
			}
			return { success: false, error: "Could not parse user identifier" };
		}

		try {
			const room = state.data?.room || (await runtime.getRoom(message.roomId));
			const serverId = room?.messageServerId;
			if (!serverId) {
				if (callback) {
					await callback?.({
						text: "I couldn't determine the current server.",
						source: "discord",
					});
				}
				return { success: false, error: "Could not determine current server" };
			}

			const guild = await discordService.client.guilds.fetch(serverId);

			let member: GuildMember | null = null;

			// Handle "self" request
			if (userInfo.userIdentifier === "self") {
				const content = message.content as {
					user_id?: string;
					userId?: string;
				};
				const authorId = content.user_id || content.userId;
				if (authorId && typeof authorId === "string") {
					const cleanId = authorId.replace("discord:", "");
					try {
						member = await guild.members.fetch(cleanId);
					} catch (_e) {
						// User not found
					}
				}
			} else {
				// Remove mention formatting if present
				const cleanIdentifier = userInfo.userIdentifier.replace(/[<@!>]/g, "");

				// Try to fetch by ID first
				if (/^\d+$/.test(cleanIdentifier)) {
					try {
						member = await guild.members.fetch(cleanIdentifier);
					} catch (_e) {
						// Not an ID or user not found
					}
				}

				// If not found by ID, search by username or display name
				if (!member) {
					const members = await guild.members.fetch();
					member =
						members.find(
							(m) =>
								m.user.username.toLowerCase() ===
									userInfo.userIdentifier.toLowerCase() ||
								m.displayName.toLowerCase() ===
									userInfo.userIdentifier.toLowerCase() ||
								(m.user.discriminator !== "0" &&
									`${m.user.username}#${m.user.discriminator}`.toLowerCase() ===
										userInfo.userIdentifier.toLowerCase()),
						) || null;
				}
			}

			if (!member) {
				if (callback) {
					await callback?.({
						text: `I couldn't find a user with the identifier "${userInfo.userIdentifier}" in this server.`,
						source: "discord",
					});
				}
				return {
					success: false,
					error: `User not found: ${userInfo.userIdentifier}`,
				};
			}

			const infoText = formatUserInfo(member, userInfo.detailed);

			const response: Content = {
				text: infoText,
				source: message.content.source,
			};

			if (callback) {
				await callback?.(response);
			}
			return { success: true, text: response.text };
		} catch (error) {
			runtime.logger.error(
				{
					src: "plugin:discord:action:get-user-info",
					agentId: runtime.agentId,
					error: error instanceof Error ? error.message : String(error),
				},
				"Error getting user info",
			);
			if (callback) {
				await callback?.({
					text: "I encountered an error while getting user information. Please try again.",
					source: "discord",
				});
			}
			return {
				success: false,
				error: error instanceof Error ? error.message : String(error),
			};
		}
	},
	examples: (spec.examples ?? []) as ActionExample[][],
};

export default getUserInfo;
